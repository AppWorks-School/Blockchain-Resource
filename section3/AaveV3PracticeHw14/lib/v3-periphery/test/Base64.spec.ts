import { ethers } from 'hardhat'
import { base64Encode } from './shared/base64'
import { expect } from './shared/expect'
import { Base64Test } from '../typechain'
import { randomBytes } from 'crypto'
import snapshotGasCost from './shared/snapshotGasCost'

function stringToHex(str: string): string {
  return `0x${Buffer.from(str, 'utf8').toString('hex')}`
}

describe('Base64', () => {
  let base64: Base64Test
  before('deploy test contract', async () => {
    base64 = (await (await ethers.getContractFactory('Base64Test')).deploy()) as Base64Test
  })

  describe('#encode', () => {
    it('is correct for empty bytes', async () => {
      expect(await base64.encode(stringToHex(''))).to.eq('')
    })

    for (const example of [
      'test string',
      'this is a test',
      'alphabet soup',
      'aLpHaBeT',
      'includes\nnewlines',
      '<some html>',
      'ðŸ˜€',
      'f',
      'fo',
      'foo',
      'foob',
      'fooba',
      'foobar',
      'this is a very long string that should cost a lot of gas to encode :)',
    ]) {
      it(`works for "${example}"`, async () => {
        expect(await base64.encode(stringToHex(example))).to.eq(base64Encode(example))
      })

      it(`gas cost of encode(${example})`, async () => {
        await snapshotGasCost(base64.getGasCostOfEncode(stringToHex(example)))
      })
    }

    describe('max size string (24kB)', () => {
      let str: string
      before(() => {
        str = Array<null>(24 * 1024)
          .fill(null)
          .map((_, i) => String.fromCharCode(i % 1024))
          .join('')
      })
      it('correctness', async () => {
        expect(await base64.encode(stringToHex(str))).to.eq(base64Encode(str))
      })
      it('gas cost', async () => {
        await snapshotGasCost(base64.getGasCostOfEncode(stringToHex(str)))
      })
    })

    it('tiny fuzzing', async () => {
      const inputs = []
      for (let i = 0; i < 100; i++) {
        inputs.push(randomBytes(Math.random() * 100))
      }

      const promises = inputs.map((input) => {
        return base64.encode(`0x${input.toString('hex')}`)
      })

      const results = await Promise.all(promises)

      for (let i = 0; i < inputs.length; i++) {
        expect(inputs[i].toString('base64')).to.eq(results[i])
      }
    }).timeout(300_000)
  })
})
