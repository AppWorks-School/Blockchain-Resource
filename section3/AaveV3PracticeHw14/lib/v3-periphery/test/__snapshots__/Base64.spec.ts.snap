// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base64 #encode gas cost of encode(<some html>) 1`] = `1497`;

exports[`Base64 #encode gas cost of encode(aLpHaBeT) 1`] = `1256`;

exports[`Base64 #encode gas cost of encode(alphabet soup) 1`] = `1727`;

exports[`Base64 #encode gas cost of encode(f) 1`] = `763`;

exports[`Base64 #encode gas cost of encode(fo) 1`] = `774`;

exports[`Base64 #encode gas cost of encode(foo) 1`] = `769`;

exports[`Base64 #encode gas cost of encode(foob) 1`] = `1004`;

exports[`Base64 #encode gas cost of encode(fooba) 1`] = `1015`;

exports[`Base64 #encode gas cost of encode(foobar) 1`] = `1010`;

exports[`Base64 #encode gas cost of encode(includes
newlines) 1`] = `1979`;

exports[`Base64 #encode gas cost of encode(test string) 1`] = `1497`;

exports[`Base64 #encode gas cost of encode(this is a test) 1`] = `1738`;

exports[`Base64 #encode gas cost of encode(this is a very long string that should cost a lot of gas to encode :)) 1`] = `6083`;

exports[`Base64 #encode gas cost of encode(ðŸ˜€) 1`] = `1004`;

exports[`Base64 #encode max size string (24kB) gas cost 1`] = `3731900`;
